src/main/UserActivityMonitor.js

```diff
@@ -0,0 +1,170 @@
+      return;
+    }
+
+    if (this.systemIdleTimeIntervalID >= 0) {
+      this.emit('error', new Error('User activity monitoring is already in progress'));
+      return;
+    }
+
+    this.config = Object.assign({}, this.config, config);
+
+    // NOTE: electron.powerMonitor cannot be referenced until the app is ready
+    electron.powerMonitor.on('suspend', this.handleSuspend);
+    electron.powerMonitor.on('resume', this.handleResume);
+    electron.powerMonitor.on('shutdown', this.handleShutdown);
+    electron.powerMonitor.on('lock-screen', this.handleLockScreen);
+    electron.powerMonitor.on('unlock-screen', this.handleUnlockScreen);
+
+    this.systemIdleTimeIntervalID = setInterval(() => {
+      try {
+        electron.powerMonitor.querySystemIdleTime((idleTime) => {
+          this.updateIdleTime(idleTime);
+        });
+      } catch (err) {
+        console.log('Error getting system idle time:', err);
+      }
+    }, this.config.regularUpdateFrequencyMs);
+  }
+
+  /**
+   * Stop monitoring system events and idle time
+   */
+  stopMonitoring() {
+    electron.powerMonitor.off('suspend', this.handleSuspend);
+    electron.powerMonitor.off('resume', this.handleResume);
+    electron.powerMonitor.off('shutdown', this.handleShutdown);
+    electron.powerMonitor.off('lock-screen', this.handleLockScreen);
+    electron.powerMonitor.off('unlock-screen', this.handleUnlockScreen);
+
+    clearInterval(this.systemIdleTimeIntervalID);
+  }
+
+  /**
+   * Updates internal idle time properties and conditionally triggers updates to user activity status
+   *
+   * @param {integer} idleTime
+   * @private
+   */
+  updateIdleTime(idleTime) {
+    this.idleTime = idleTime;
+
+    if (this.idleTime > this.config.userActivityTimeoutSec) {
+      this.updateUserActivityStatus(false);
+    } else if (!this.forceUserIsInactive && this.idleTime < this.config.userActivityTimeoutSec) {
+      this.updateUserActivityStatus(true);
+    }
+  }
+
+  /**
+   * Updates user activity status if changed and triggers a status update
+   *
+   * @param {boolean} isActive
+   * @private
+   */
+  updateUserActivityStatus(isActive = false) {
+    if (isActive !== this.isActive) {
+      this.isActive = isActive;
+      this.sendStatusUpdate();
+    } else if (new Date().getTime() - this.lastStatusUpdate > this.config.statusUpdateThresholdMs) {
+      this.sendStatusUpdate();
+    }
+  }
+
+  /**
+   * Sends an update with user activity status and current system idle time
+   *
+   * @emits {status} emitted at regular, definable intervals providing an update on user active status and idle time
+   * @private
+   */
+  sendStatusUpdate() {
+    this.lastStatusUpdate = new Date().getTime();
+    this.emit('status', {
+      userIsActive: this.isActive,
+      idleTime: this.idleTime,
+    });
+  }
+
+  /**
+   * System event handlers
+   *
+   * @private
+   */
+  handleSuspend() {
```

Comment:
It seems like there's only 2 different behaviours here. Would it make sense to just have two methods that are reused for different events? 0/5